숫자게임 [STEP2] 애종, LJ

limjy0 opened the pull request:
안녕하세요 바드(@bar-d) 
즐거운 토요일 오후 보내고 계신가요? 햇살좋은날 이렇게 양심없이 PR보냅니다😅

STEP 2
 ### 고민했던 점
 - 사용자 정의타입을 사용하지 않고 , 전역변수의 사용을 최소화 하려고 했습니다. 모든 기능을 함수로 표현하다보니 함수의 반환값을 많이 활용하게 된 것 같습니다.
- 예를 들어 checkError 혹은 judgeVictory 와 같이 에러가 존재하는지 안하는지, 사용자의 승리인지 컴퓨터의 승리인지를 함수의 반환값으로 상위함수에 알려주어 프로그램의 진행방향을 결정했습니다.

- 아무래도 가장 시간이 오래걸렸던 부분은 Optional, 오류 처리 였습니다. 사전과제 떄와 비슷하면서도 달랐던 것 같습니다.

- 3개의 숫자를 입력받아 `Array`에 3가지 요소를 담았기 때문에 하나하나씩 제대로 입력했는지 검증하는 과정을 설계하는게 고민이 많았습니다. 고차함수 `map`, `compactMap`, `filter`등을 사용해 count를 통해 조건을 만족시키고자 했습니다.

- 구성 요소중 nil 을 제외하는 `compactMap`을 사용하면 [String?] 과 같이 배열 내부에 있는 `Optional`을 오류없이 unwrap된다는걸 알았습니다.

- checkError함수 안에서 Int형배열로 변환한 배열 convertToNumbers을 다른 함수에서 사용할 때, 전역변수를 사용하지 않고 사용할 방법을 고민하였습니다. 


 ---

 ### 조언을 얻고 싶은 부분
 - xcode 내에서 함수 혹은 클래스 변수의 순서를 어떤 기준으로 써내려가는게 좋은지 궁금합니다. 저희는 아래에 startGame()을 두고 startGame()내부의 진행순서에 따라 역순으로 함수를 배치했는데, 또 추천해주시거나 컨벤션같은게 있는지 궁금합니다.

- 변수선언을 여러개를 하는 것과 조건문 자체에 조건대상이되는 수식을 넣는것 중 선택할 때의 기준(간결성, 명확성 등)이 있을까요?
- let a = 10
let b = 10
let c = a + b
if c > 100 { ... } //변수선언을 여러개를 하는 것
- let a = 10
let b = 10
if a + b 100 { ... } //조건문 자체에 조건대상이되는 수식을 넣는것

- getInputNumbers함수에서 `optional binding`한 배열을 반환값으로 정하였습니다. `optional binding`실패 시 임의로 정한 값을 반환하여 사용하는것이 맞는지 궁금합니다.
--- 

바드 너무 감사했습니다. 항상 실시간으로 함께 봐주시고 친절히 상세하게 알려주신 덕분에 효율적으로 진행할 수 있었습니다. 주말시간을 빼앗아서 미안하고 감사합니다!!!

limjy0 committed:
feat: STEP2 함수 추가
- 함수이름 변경 startGame -> playGame
- 함수이름 변경 printRandomNumberArray -> printRandomNumbers
- startGame 함수 재정의
- selectMenu 함수 추가
- showMenu 함수 추가
- selectedMenu 변수 추가

jonghancha committed:
test: test branch 생성

limjy0 committed:
feat: STEP2 함수 추가 및 refactoring
- getInputNumbers 함수 , 검증 포함된 부분 checkError함수로 분리
- checkError 함수 추가
- showMenu 함수 추가
- printRemainTryCount 함수 추가, 컴퓨터 승리시 남은 기회 출력
- judgeVictory 함수 추가, 반환값 생성
- printInputCondition 함수 추가, 출력메시지 변경

limjy0 committed:
feat: STEP2 함수 추가, 함수 반환값 생성
- getInputNumbers 함수 수정, 검증 포함된 부분 checkError함수로 분리
- checkError 함수 추가
- printRemainTryCount 함수 추가, 컴퓨터 승리시 남은 기회 출력
- judgeVictory 함수, 반환값 생성
- printInputCondition 함수, 출력메시지 변경

bar-d commented:
주말에도 STEP2 진행하시느라 고생하셨습니다.  

## 👏칭찬드리고 싶은 점

- 해당 프로젝트 명세서에 명시된 전역변수만 선언하신 점👏  
기능 단위별로 최대한 커밋할려고 노력해주신점 좋습니다
- 이전보다 코드 가독성이 좋아졌어요 👏👏  
컨벤션도 잘 지키시려고 노력하시고 이전보다 훨씬 코드를 읽기 편해졌네요  
여러분도 느껴지시나요?

## ⚠️프로젝트 명세서를 다시 확인해주세요
- 실행 화면이 프로젝트 요구사항과는 조금 다른 것 같아요   
  개행을 다시 한번 확인해주세요  
![Screen Shot 2022-08-20 at 6 51 41 PM](https://user-images.githubusercontent.com/92622931/185740003-35e5dc9a-6ed6-4fc1-95f6-a11eac3f87f6.png)


## 커밋 단위를 좀 더 세분화해주세요
- 하나의 커밋이 포함하고 있는 내용이 너무 많은 것 같아요  
본문의 내용들을 하나의 커밋들로 나눌 수 있을 것 같아요

## 고민했던 점

>사용자 정의타입을 사용하지 않고 , 전역변수의 사용을 최소화 하려고 했습니다.  
 모든 기능을 함수로 표현하다보니 함수의 반환값을 많이 활용하게 된 것 같습니다.

- 함수의 반환값이 있다고 안좋은 코드가 아닙니다  
적절하게 반환값을 사용하셨다면 좋은 코드가 아닐까요?

>예를 들어 checkError 혹은 judgeVictory 와 같이 에러가 존재하는지 안하는지, 사용자의 승리인지 컴퓨터의 승리인지를 함수의 반환값으로 상위함수에 알려주어 프로그램의 진행방향을 결정했습니다.

- 매개변수를 넣어 반환해준 점 좋습니다

>아무래도 가장 시간이 오래걸렸던 부분은 Optional, 오류 처리 였습니다.  
 사전과제 떄와 비슷하면서도 달랐던 것 같습니다.

- 사전과제와 어떻게 달랐나요?  
어떤 부분에서 어떠한 오류처리를 해주셨나요?

>3개의 숫자를 입력받아 Array에 3가지 요소를 담았기 때문에 하나하나씩 제대로 입력했는지 검증하는 과정을 설계하는게 고민이 많았습니다.   
고차함수 map, compactMap, filter등을 사용해 count를 통해 조건을 만족시키고자 했습니다.

- 고차함수를 사용해주신점 좋습니다  
`map`, `compactMap`, `filter` 는 어떠한 역할을 해주는 고차함수인가요?

>구성 요소중 nil 을 제외하는 compactMap을 사용하면 [String?] 과 같이 배열 내부에 있는 Optional을 오류없이 unwrap된다는걸 알았습니다.

- [해당 문서](https://developer.apple.com/documentation/swift/sequence/compactmap(_:))를 읽고 이유도 적어주시면 좋아요

>checkError함수 안에서 Int형배열로 변환한 배열 convertToNumbers을 다른 함수에서 사용할 때, 전역변수를 사용하지 않고 사용할 방법을 고민하였습니다.
- 고민하신 이유와 해결한 방법을 같이 써주시면 좋을 것 같습니다

## 조언을 얻고 싶은 부분

>xcode 내에서 함수 혹은 클래스 변수의 순서를 어떤 기준으로 써내려가는게 좋은지 궁금합니다.   
저희는 아래에 startGame()을 두고 startGame()내부의 진행순서에 따라 역순으로 함수를 배치했는데, 또 추천해주시거나 컨벤션같은게 있는지 궁금합니다.

- 정답이 있는 것은 아니지만, 저는 프로퍼티, 함수순으로 두는 편이고 호출순서대로 함수를 위에서 아래로 배치합니다  
코드를 읽을 때 위에서 아래로 읽으니까요

>변수선언을 여러개를 하는 것과 조건문 자체에 조건대상이되는 수식을 넣는것 중 선택할 때의 기준(간결성, 명확성 등)이 있을까요?
```
let a = 10
let b = 10
let c = a + b
if c > 100 { ... } //변수선언을 여러개를 하는 것

let a = 10
let b = 10
if a + b > 100 { ... } //조건문 자체에 조건대상이되는 수식을 넣는것
```

- 상황에 따라 다를 것 같아요  
저라면 현재 예시처럼 a와 b를 그저 더한 값이라면 상수를 선언해주지 않고 아래처럼 할 것 같네요  
현재 예시에서는 프로퍼티를 따로 선언할 필요는 없다고 생각합니다  

>getInputNumbers함수에서 optional binding한 배열을 반환값으로 정하였습니다. optional binding실패 시 임의로 정한 값을 반환하여 사용하는것이 맞는지 궁금합니다.
- 좋지 않은 방법인것 같아요 `guard let`을 사용하면 코드를 개선할 수 있을 것 같아요

bar-d commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -3,73 +3,122 @@
-
 import Foundation
 
 var computerNumberArray: Array<Int> = []
 var tryCount: Int = 9
```
프로젝트에 명시된 전역변수만 사용한 점 좋아요

bar-d commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -3,73 +3,122 @@
-}
-
-func judgeVictory(by strikeCount: Int){
+func judgeVictory(by strikeCount: Int) -> String {
     if strikeCount == 3 {
         print("사용자 승리...!")
+        return "win"
     } else if tryCount == 0 {
         print("컴퓨터 승리...!")
     }
+    
+    return "lose"
 }
```
return 값을 넣어준 이유가 무엇일까요?  
필요한 반환값일까요?

jylim72 replied:
return값을 넣어준 이유
사용자가 승리할 경우에는 남은기회를 출력하지 않고 새로 게임을 시작하기 위해서 넣었습니다.
> if judgeVictory(by: strikeAndBall[0]) == "win" { return }

judgeVictory 함수호출 코드와 같이 if문으로 사용자가 승리할 경우(win) return값을 이용하여 playGame함수의 while문을 빠져나가 다시 새로운게임을 시작하거나 게임종료를 선택하도록 했습니다.
# 
필요한 return값인지 고민해보았습니다.
return값 없이도 사용자가 승리하고, 남은기회 출력하지않고, 새로운 게임을 시작하는 방법이 있을지 생각했습니다.
남은기회 출력여부는 printRemainTryCount함수가 결정한다는 점을 상기했습니다. 
남은기회 변수인 tryCount값에 따라 출력여부가 결정되고,
사용자가 승리할 경우에는 tryCount를 0으로 만들어 남은기회를 출력하지 않고 새로 게임을 시작는 것으로 해결할 수 있었습니다.
#
- judgeVictory 함수 refactoring

수정 후
```swift
func judgeVictory(by strikeCount: Int) {
    if strikeCount == 3 {
        print("사용자 승리...!")
    } else if tryCount == 0 {
        print("컴퓨터 승리...!")
    }
}
```
#
- judgeVictory함수 호출하는 코드 refactoring

수정 전
```swift
if judgeVictory(by: strikeAndBall[0]) == "win" { return }
```

수정 후
```swift
judgeVictory(by: strikeAndBall[0])
```
#
- printRemainTryCount 함수 refactoring

수정 전
```swift
func printRemainTryCount(strikeCount: Int) {
    if strikeCount != 3 && tryCount > 0 {
        print("남은 기회 : \(tryCount)")
    }
}
```

수정 후
```swift
func printRemainTryCount(strikeCount: Int) {
    if strikeCount != 3 && tryCount > 0 {
        print("남은 기회 : \(tryCount)")
    } else {
        tryCount = 0
    }
}
```

bar-d replied:
훨씬 코드가 깔끔해졌네요 👍

bar-d commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -3,73 +3,122 @@
+    let convertToNumbers: Array<Int> = input.map { Int($0) }.compactMap { $0 }
+    guard input[0] != "", Set(input).count == 3,
+          convertToNumbers.count == 3,
+          convertToNumbers.filter { 1...9 ~= $0 }.count == 3 else { return false }
```
![Screen Shot 2022-08-20 at 10 46 17 PM](https://user-images.githubusercontent.com/92622931/185749043-fc83f391-c676-44e3-ab0e-225e18669302.png)
해당 경고창이 뜨는 이유는 무엇일까요?  
경고창이 안뜨도록 변경해주세요

jylim72 replied:
수정 후
```swift
let convertToNumbers: Array<Int> = input.map { Int($0) }.compactMap { $0 }
    guard input[0] != "", Set(input).count == 3,
          convertToNumbers.count == 3,
          convertToNumbers.filter({ 1...9 ~= $0 }).count == 3 else { return false }
```
{ 1...9 ~= $0 } 이 클로저가 구문가운데에 있어 guard문 중간에 혼동될 수 있으니 소괄호로 감싸서 인수임을 명확하게 나타내라는 경고창 메세지대로 수정하였습니다.

bar-d replied:
앞으로도 경고창은 뜨지 않도록 노력해주세요 👍

bar-d commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -3,73 +3,122 @@
+    return true
+}
+
+func getInputNumbers() -> Array<String> {
+    if let input: Array<String> = readLine()?.components(separatedBy: " ") {
+        return input
     }
+    
+    return ["nil input"]
+}
```
해당 함수가 어떤 역할을 하는지 생각해보시고 리팩토링 해주세요  
임의적으로 반환값을 넣어주는 것이 좋은 방법처럼 보이지는 않아요

jylim72 replied:
수정 후
```swift
func getInputNumbers() -> String? {
    guard let input: String = readLine() else {
        return nil
    }
    return input
}
```

이 함수의 역할은 사용자로부터 값을 입력받는 것 입니다.
생각해보니 ["nil input"] 은 사용자가 입력한 값이 아니므로 근거가 없어보여 리펙토링 했습니다.
    
그래서 사용자로부터 입력받은 값을 공백으로 분리 후 배열로 반환하는
> components(separatedBy: " ")

이 코드는 checkError함수(사용자입력값 유효성검증)로 옮겼습니다.

bar-d commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -3,73 +3,122 @@
+
+func selectMenu() -> String {
+    showMenu()
+    if let input = readLine() {
+        return input
+    }
+    
+    return " "
```
`if let`과 `guard let`의 차이점은 무엇일까요?

jylim72 replied:
수정 후
```swift
guard let input = readLine() else {
       return ""
    }

    return input
```
* optional binding

|          | if let | guard let |
| -------- | -------- | -------- |
|        | nil이면 어떻게 처리하고, nil이 아니면 그 결과에 따라 어떻게 처리할 지 각각 정하고 싶을 때 사용됨 | 예외 여부를 처리하기 위해 사용됨 |
| binding한 상수 scope | if블록 내부O,외부X| guard블록 내부X,외부O |

이 구문의 목적은 사용자가 입력한 값을 optional binding하는 것입니다. 
binding이 안되는 예외 하나만 처리하면 되기때문에 guard문을 사용하는 것이 더 낫다고 이해하면 될까요🤔

bar-d replied:
둘 중에 `if let`을 사용하신 이유가 궁금해서 물어봤어요
차이점을 잘 알고 계시네요 👍

bar-d commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -3,73 +3,122 @@
+    var selectedMenu: String
+    repeat {
+        selectedMenu = selectMenu()
+        switch selectedMenu {
+            case "1":
+                playGame()
+                tryCount = 9
+            case "2":
+                break
+            default:
+                print("입력이 잘못되었습니다.\n")
```
switch case 구문활용 좋습니다👍

limjy0 committed:
feat: 함수 refactoring
- 함수 호출순서대로 코드순서 변경
- getInputNumbers 함수 반환타입 Bool에서 Array<Int>? 로 변경
- playGame 함수의 inputNumbers상수 타입 변경
- checkError 함수 입력값 배열타입으로 변환하는 역할 추가

limjy0 committed:
feat: 남은기회 출력 refactoring
- 남은기회 0일 경우 출력 안되도록 수정

limjy0 committed:
feat: 함수 refactoring, 주석 제거
- judgeVictory 함수 return값 삭제
- printRemainTryCount 함수 else문 추가
- 주석 제거

limjy0 commented:
> 아무래도 가장 시간이 오래걸렸던 부분은 Optional, 오류 처리 였습니다. 사전과제 떄와 비슷하면서도 달랐던 것 같습니다.
> ㄴ 사전과제와 어떻게 달랐나요? 어떤 부분에서 어떠한 오류처리를 해주셨나요?

| 달랐던 부분   | 사전과제 | 숫자게임 |
| -------- | -------- | -------- |
| optional | 강제추출 여러번 사용 | 안전한 옵셔널바인딩 사용 |
| 오류처리: 사용자입력값 검증 | 이중 if문 사용 | guard문과 함수를 이용하여 검증 |

이 외에도 상수/변수, 전역변수/지역변수를 한번 더 생각해보고 코드를 작성한 점이 달랐습니다.

---
> 3개의 숫자를 입력받아 Array에 3가지 요소를 담았기 때문에 하나하나씩 제대로 입력했는지 검증하는 과정을 설계하는게 고민이 많았습니다.
고차함수 map, compactMap, filter등을 사용해 count를 통해 조건을 만족시키고자 했습니다.
> ㄴ map, compactMap, filter 는 어떠한 역할을 해주는 고차함수인가요?
* map : 자신을 호출할 때 매개변수로 전달된 함수를 실행하여 그 결과를 다시 반환해주는 함수, 기존 데이터를 변형하는데 많이 사용함
* compactMap : 각 요소를 변환하였을 때 non-nil 결과값들로만 이루어진 배열을 리턴하는 함수
* filter : 컨테이너 내부의 값을 특정조건에 맞게 걸러서 추출하는 역할을 하는 함수, 매개변수로 전달되는 함수의 반환타입은 Bool, true인 값을 추출함 


참고 : 링크주신 [해당문서](https://developer.apple.com/documentation/swift/sequence/compactmap(_:)) , 야곰책 스위프트 프로그래밍

---
> 구성 요소중 nil 을 제외하는 compactMap을 사용하면 [String?] 과 같이 배열 내부에 있는 Optional을 오류없이 unwrap된다는걸 알았습니다.
ㄴ 해당 문서를 읽고 이유도 적어주시면 좋아요


> 
```swift
func compactMap<ElementOfResult>(_ transform: (Self.Element) throws -> ElementOfResult?) rethrows -> [ElementOfResult]
```
transform클로저가 자기자신의 요소를 받아서 던지고, 받은요소에서 옵셔널값을 다시 던져서 배열로 return하는 함수
ex) 원본배열을 한 번 순회해서 결과 배열을 만들고, //[1, 2, nil, 4, 5]
이를 다시 순회해서 옵셔널값을 제거한다. //[1, 2, 4, 5]
결과만 보았을 때 옵셔널값이 제거되어 나오기 때문에 배열 안에 옵셔널값이 있어도 오류없이 unwrap한다는 것을 알 수 있었습니다.

참고 : 링크주신 [해당문서](https://developer.apple.com/documentation/swift/sequence/compactmap(_:)), [사이트intrepidgeeks](https://intrepidgeeks.com/tutorial/swift-compactmap-flatmap) 

---
> checkError함수 안에서 Int형배열로 변환한 배열 convertToNumbers을 다른 함수에서 사용할 때, 전역변수를 사용하지 않고 사용할 방법을 고민하였습니다.
ㄴ 고민하신 이유와 해결한 방법을 같이 써주시면 좋을 것 같습니다
* 사용자로부터 입력받은 숫자배열은 두 곳에서 사용해야 했습니다.
1. 입력받은 숫자의 유효성 검증
   이 경우 같은 checkError함수 안에서 사용하므로 어려움이 없었습니다.
2. 컴퓨터의 숫자와 비교
    다른 함수 checkStrikeAndBall에서 숫자배열을 사용해야 했습니다.
    return값으로 숫자배열을 넘겨주어  checkStrikeAndBall함수의 매개변수로 받아와 사용하여 해결하였습니다.

limjy0 commented:
꼼꼼하게 피드백 주신 덕분에 많이 배운 한 주였습니다.🤓 고생 많으셨습니다.👏👏👏
🎉바드(@bar-d )🎉감사합니다!!!

bar-d commented:
1주차 프로젝트 하시느라 고생하셨습니다
